File: solutions/java/bank-account/1/src/main/java/BankAccount.java
class BankAccount {

    private int balance;
    private boolean isClosed = true;

    void open() throws BankAccountActionInvalidException {
        if (!isClosed) throw new BankAccountActionInvalidException("Account already open");

        balance = 0;
        isClosed = false;
    }

    void close() throws BankAccountActionInvalidException {
        if (isClosed) throw new BankAccountActionInvalidException("Account not open");

        isClosed = true;
    }

    synchronized int getBalance() throws BankAccountActionInvalidException {
        checkIsOpen();

        return balance;
    }

    synchronized void deposit(int amount) throws BankAccountActionInvalidException {
        checkIsOpen();
        checkAmount(amount);

        balance += amount;
    }

    synchronized void withdraw(int amount) throws BankAccountActionInvalidException {
        checkIsOpen();
        checkAmount(amount);
        checkBalance(amount);

        balance -= amount;
    }

    private void checkIsOpen() throws BankAccountActionInvalidException {
        if (isClosed) throw new BankAccountActionInvalidException("Account closed");
    }

    private void checkAmount(int amount) throws BankAccountActionInvalidException {
        if (amount < 0) throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");
    }


    private void checkBalance(int amount) throws BankAccountActionInvalidException {
        if (balance < amount) throw new BankAccountActionInvalidException("Cannot withdraw more money than is currently in the account");
    }

}

File: solutions/lua/bank-account/1/bank-account.lua
local BankAccount = { _balance = 0, _closed = false }
BankAccount.__index = BankAccount

function BankAccount:new()
    local instance = setmetatable({}, BankAccount)
    return instance
end

function BankAccount:balance()
    return self._balance
end

function BankAccount:close()
    self._closed = true
end

function BankAccount:deposit(amount)
    if self._closed then
        error("Account is closed")
    end
    if amount <= 0 then
        error("Deposits must Be positive numbers")
    end

    self._balance = self._balance + amount
end

function BankAccount:withdraw(amount)
    if self._closed then
        error("Account is closed")
    end
    if amount <= 0 then
        error("Withdrawals must be positive numbers")
    end
    if amount > self._balance then
        error("Insuficient balance for withdrawal")
    end

    self._balance = self._balance - amount
end

return BankAccount
